### PR TITLE
Added @ts-expect-error to @ts-ignore directives

### LINKED ISSUE #29394: ts-ignore should error if next line is not an error
http://www.typescriptlang.org/play/#src=%2F%2F%20%40ts-ignore%0D%0Aconst%20a%3A%20number%20%3D%205%3B

In the example above, the `@ts-ignore` line should error because there is no TypeScript error to be ignored. This is the same behavior as Flow's `suppress_comment` [feature](https://flow.org/en/docs/config/options/#toc-suppress-comment-regex).

Alternatively, if you want to maintain this behavior, consider creating a `@ts-expect-error`.

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Similar to `// @ts-ignore`, but will itself cause a new error diagnostic if it does not cause an existing diagnostic to be ignored.

Technical summary:
1. The scanner will now keep track of `CommentDirective`s it comes across: both `@ts-expect-error` and `@ts-ignore`
2. During type checking, the program will turn those directives into a map keying them by line number
3. For each diagnostic, if it's preceded by a directive, that directive is marked as "used"
4. All `@ts-expect-error` directives not marked as used generate a new diagnostic error

Fixes #29394

### CODE REVIEW

**Comment by orta** [comment]
I'd like to try this in a playground:

@typescript-bot perf test
@typescript-bot pack this

**Comment by JoshuaKGoldberg** [comment]
Re that On-Demand Benchmark test failure: Monaco _check_ time went up 7.14% ([line link](https://typescript.visualstudio.com/TypeScript/_build/results?buildId=60755&view=logs&j=fd490c07-0b22-5182-fac9-6d67fe1e939b&t=a12e1e51-0647-5e2d-1f8a-0f2707573119&l=132)).  but _emit_ time decreased by almost as much (31s out of 36s), and the _total_ time went up only by 0.57%: 

> ```
> │ Monaco - node (v8.9.0, x86)                                                                     │
> ├─────────────┬────────────────────┬────────────────────┬───────────────────┬──────────┬──────────┤
> │ Memory used │ 203,369k (± 0.02%) │ 204,511k (± 0.01%) │ +1,141k (+ 0.56%) │ 204,463k │ 204,576k │
> │ Parse Time  │    1.60s (± 0.35%) │    1.61s (± 0.72%) │  +0.01s (+ 0.56%) │    1.59s │    1.63s │
> │ Bind Time   │    0.76s (± 1.35%) │    0.76s (± 0.77%) │  -0.00s (- 0.26%) │    0.74s │    0.77s │
> │ Check Time  │    5.08s (± 1.78%) │    5.45s (± 1.22%) │  +0.36s (+ 7.14%) │    5.22s │    5.57s │
> │ Emit Time   │    3.21s (± 2.34%) │    2.90s (± 3.30%) │  -0.31s (- 9.66%) │    2.83s │    3.28s │
> ├─────────────┼────────────────────┼────────────────────┼───────────────────┼──────────┼──────────┤
> │ Total Time  │   10.65s (± 0.33%) │   10.71s (± 0.56%) │  +0.06s (+ 0.57%) │   10.59s │   10.87s │
> ```

...so this feels like it's not a serious performance degradation, and this PR is still valid? 😁

**Comment by davidgomes** [comment]
@JoshuaKGoldberg I've also been toying around with implementing this. However, I was making `@ts-ignore` and `@ts-expect-error` pragmas instead of adding a new concept (Comment Directive).

I added the `@ts-expect-error` as a new pragma:

```
        "ts-expect-error": {
            kind: PragmaKindFlags.AnySingleLine
        }
```

And then I parse it in `processCommentPragmas` in `parser.ts`:

```
        iterateAllCommentRanges(sourceText, (range) => {
            const comment = sourceText.substring(range.pos, range.end);
            extractPragmas(pragmas, range, comment, PragmaKindFlags.AnySingleLine);
        });
```

(`iterateAllCommentRanges` is a new function I added to `scanner.ts`)

While I couldn't find a formal definition of "pragmas" in programming languages, in some programming languages, a "pragma" can be in places other than just the top of the file (https://dlang.org/spec/pragma.html).

Here's my in progress diff: https://gist.github.com/davidgomes/5cb71342b6d2cf95ffa05a767eb802c7

**Comment by rauschma** [comment]
I’d love to have something more powerful:

```ts
//@ts-expect-error 2322 "Type 'number' is not assignable to type 'never'."
```

Maybe with the option to omit trailing characters of the message and/or the error number. The former could be achieved via regular expressions:

```ts
//@ts-expect-error 2322 /^Type 'number' is not assignable to type 'never'\.$/
```

Rationale: It’s not only a problem to expect errors where there aren’t any, but also to expect the wrong kind of error.

**Comment by orta** [comment]
Re-running perf tests, I'm not really convinced this could add 7% after reading the PR

@typescript-bot perf test
@typescript-bot pack this

**Comment by orta** [comment]
These numbers make more sense to me - basically everything is varying in the 0.x range on a plus or minus which is basically just normal noise in perf tests.

File: src/compiler/program.ts [hunk 1]
```diff
@@ -1749,49 +1746,68 @@ namespace ts {
...
+            const { diagnostics, directives } = getDiagnosticsPastDirectives(sourceFile, sourceFile.commentDirectives, flatDiagnostics);
+
+            for (const errorExpectation of directives.getUnusedExpectations()) {
+                diagnostics.push(createDiagnosticForRange(sourceFile, errorExpectation.range, Diagnostics.Unused_ts_expect_error_directive));
+            }
+
+            return diagnostics;
+        }
+
+        function getDiagnosticsPastDirectives(sourceFile: SourceFile, commentDirectives: CommentDirective[], flatDiagnostics: Diagnostic[]) {
```
Comment (orta) [inline src/compiler/program.ts]: My only feedback is that I initially didn't grok what this function did by the name, but after I got it, I couldn't come up with a more satisfying one.

Perhaps a JSDoc comment above would be enough?

Comment (JoshuaKGoldberg) [reply src/compiler/program.ts]: Yeah, makes sense. I also struggled with this one... 😄 

Also, slight amusement:

Screenshot of "my comment above" actually referring to text below

**Comment by ExE-Boss** [comment]
This would make it possible to deprecate `$ExpectError` in **dtslint**.

File: src/compiler/program.ts (hunk 1 above)
Comment (orta) [reply src/compiler/program.ts]: OK, maybe `getDiagnosticsWithPrecedingDirectives` ?

**Comment by orta** [comment]
OK, this looks good to go - thanks!

I don't know if it's in TypeScript, or in vscode, but we should also get `// @ts-expect-error` to show up in the autocomplete in comments too, but I'm good for that to be another PR. 

The playground doesn't show any of those options.

**Comment by JoshuaKGoldberg** [comment]
:+1: It's in vscode, in a list of directives: https://github.com/microsoft/vscode/pull/92093

**Comment by eps1lon** [comment]
I love this! Does this mean we can use `@ts-expect` instead of the tslint expect rule and `$ExpectError`?

**Comment by orta** [comment]
I think so, yep

**Comment by tyler1205** [comment]
As someone else already mentioned, I second that it would be nice to be able to specify exactly what error to expect (by its id/code etc). I've only used ts-ignore a few times but have already ran into the problem where I think I'm ignoring one error just to have it transform into another completely unrelated error however, unknowingly, since this transformation happened in "the meantime" while the ts-ignore directive was still active/in-place the whole time. So the new error that the old error "transformed into" went completely unnoticed from what I remember (until I removed the directive of course, which was when I found out about the change). Now I try to just avoid ts-ignore altogether because of such potential situations that may arise because it's just not worth taking that kind of a risk imo.

**Comment by asktree** [comment]
related to @tyler1205 's comment I would like to be able to specify number of errors, and that might be easier to handle.

**Comment by Asaf-S** [comment]
> related to @tyler1205 's comment I would like to be able to specify number of errors, and that might be easier to handle.

I agree, this is really needed as you can never know if you are blocking more than 1 error, or the wrong kind of error

**Pull request merged**